community/common/src/main/java/org/neo4j/kernel/lifecycle/SafeLifecycle.java

```diff
@@ -0,0 +1,180 @@
+ *  ---------------------------------------------------
+ * | PRE       |   PRE       -       -       HALT(*)   |
+ * | IDLE      |    -       IDLE   IDLE(*)    HALT     |
+ * | RUN       |    -        -      IDLE       -       |
+ * | HALT      |    -        -       -         -       |
+ *  ---------------------------------------------------
+ * </pre>
+ * A few notes:
+ * <ul>
+ * <li>will not invoke stop0() if start0() wasn't successful</li>
+ * <li>will not invoke shutdown0() if init0() wasn't successful</li>
+ * </ul>
+ * The expectation with regards to error handling and cleanup is that
+ * an unclean start() is cleaned up by the start0() method and thus
+ * the component is left in IDLE. The same goes for issues happening
+ * during init0(), which leaves the component in PRE.
+ * <p>
+ * Because of the way that {@link LifeSupport} operates today, this
+ * class will ignore stop() calls made while in IDLE. Similarly, calls
+ * to shutdown() will be ignored while in PRE. This allows this class
+ * to be managed by a {@link LifeSupport} without throwing
+ * {@link IllegalStateException} on those state transitions, which
+ * otherwise would have been disallowed and handled in the same way
+ * as other illegal state transitions.
+ * <p>
+ * This adapter will not allow a shutdown lifecycle to be reinitialized
+ * and started again.
+ */
+public abstract class SafeLifecycle implements Lifecycle
+{
+    private State state = PRE;
+
+    /**
+     * @param expected The expected from state.
+     * @param to The to state.
+     * @param op The state transition operation.
+     * @param force Causes the state to be updated regardless of if the operation throws.
+     * @throws Throwable Issues.
+     */
+    private void transition( State expected, State to, Operation op, boolean force ) throws Throwable
+    {
+        if ( state != expected )
+        {
+            throw new IllegalStateException( String.format( "Expected %s but was %s", expected, state ) );
+        }
+
+        if ( force )
+        {
+            state = to;
+            op.run();
+        }
+        else
+        {
+            op.run();
+            state = to;
+        }
+    }
+
+    @Override
+    public final synchronized void init() throws Throwable
+    {
+        transition( PRE, IDLE, this::init0, false );
+    }
+
+    @Override
+    public final synchronized void start() throws Throwable
+    {
+        transition( IDLE, RUN, this::start0, false );
+    }
+
+    @Override
+    public final synchronized void stop() throws Throwable
+    {
+        if ( state == IDLE )
+        {
+            return;
+        }
+        transition( RUN, IDLE, this::stop0, true );
+    }
+
+    @Override
+    public final synchronized void shutdown() throws Throwable
+    {
+        if ( state == PRE )
+        {
+            state = HALT;
+            return;
+        }
+        transition( IDLE, HALT, this::shutdown0, true );
+    }
+
+    public void init0() throws Throwable
```

Comment:
Are these methods (`init0`, `start0` etc...) supposed to be concrete with no-op implementations? There seems to be a change in semantics vs `LifeCycle` if so, as that interface doesn't provide any default methods or similar. Apologies if I have misunderstood the doc block up top.

Reply:
Yes. You do not have to implement any of them. It is quite common for example to only implement start() and stop(). This is akin to LifecycleAdapter. I could definitely change that if you think it is bad design and inviting errors.

Reply:
Well I was toying with the idea of inner interfaces for each state, rather than the enum (i.e. each interface represents a state, SafeLifecycle implements all of the interfaces and each method returns the interface type of the next state). This means that proper transitions would be compiler enforced, rather than checked for with the possibility of exceptions. But it might overcomplicate for little gain and I'm not sure its compatible with the idea of force (at least not simply). 

With the no-op methods. Perhaps if it is common to implement just start0 and stop0, then those should remain abstract? I assume you would never want to **not** implement those methods?

Reply:
This still has to uphold Lifecycle to be handled by LifeSupport, so not sure if it can be done the way you describe. However, if you feel like experimenting some day feel free!